### ISSUE TITLE
I don't understand these predictions

### ISSUE THREAD

**Issue by CBrauer** [dotnet/machinelearning#4168]
Hey, 

I'm trying to compute the probabilities of the predictions on my dataset.
I am using the 1.3.1 code at: https://docs.microsoft.com/en-us/dotnet/api/microsoft.ml.binaryclassificationcatalog.calibratorscatalog.platt?view=ml-dotnet
My modified code is as follows:

```
    public static void Run() {
      var mlContext = new MLContext(0);
      var dataPoints = GetDataPoints(@"/HedgeTools/Datasets/rocket-train-classify.csv");
      var trainingData = mlContext.Data.LoadFromEnumerable(dataPoints);

      var testDataPoints = GetDataPoints(@"/HedgeTools/Datasets/rocket-test-classify.csv");
      var testData = mlContext.Data.LoadFromEnumerable(testDataPoints);

      var options = new FastTreeBinaryTrainer.Options {
        EarlyStoppingMetric = EarlyStoppingMetric.L2Norm, // Use L2Norm for early stopping.
        FeatureFirstUsePenalty = 0.1,  // Create a simpler model by penalizing usage of new features.
        NumberOfTrees = 50
      };

      var pipeline = mlContext.BinaryClassification.Trainers.FastTree(options);

      var model = pipeline.Fit(trainingData);
      var transformedTestData = model.Transform(testData);
      var outScores = mlContext.Data.CreateEnumerable<ScoreValue>(transformedTestData,
                                                                  reuseRowObject: false);
      var calibratorEstimator = mlContext.BinaryClassification.Calibrators.Platt();

      // Convert IDataView object to a list.
      var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, false).ToList();
      var calibratorTransformer = calibratorEstimator.Fit(transformedTestData);
      var finalData = calibratorTransformer.Transform(transformedTestData);
      var outScoresAndProbabilities = mlContext.Data.CreateEnumerable<ScoreAndProbabilityValue>(finalData, false);

      Console.WriteLine("\nFirst 10 acutals that are true");
      Console.WriteLine("Actual Predicted     score   probability");
      Console.WriteLine("------ --------- ----------  -----------");
      var loop = 1;
      for (var index = 0; index < predictions.Count(); index++) {
        var p1 = predictions.ElementAt(index);
        if (!p1.Label) continue;
        var p2 = outScoresAndProbabilities.ElementAt(index);
        Console.WriteLine("{0, 6} {1, 9} {2, 10}  {3, 11}", p1.Label, p1.PredictedLabel, p2.Score, p2.Probability);
        if (++loop > 10) break;
      }

      var metrics = mlContext.BinaryClassification.Evaluate(transformedTestData);
      PrintMetrics(metrics);
    }
```

The results I get are as follows:

```
First 10 actuals that are true
Actual Predicted     score   probability
------ --------- ----------  -----------
  True     False  -6.288363   0.06514609
  True     False  -7.417452   0.03533126
  True     False  -6.883083    0.0473095
  True      True   2.404708    0.9079463
  True     False  -5.611444   0.09295245
  True      True  0.2836371    0.7465712
  True     False   -1.93054    0.4548629
  True     False  -3.087828    0.3014578
  True      True  0.2553135    0.7435061
  True     False  -5.725763   0.08760489

Accuracy.............0.912087912087912
AUC..................0.748737797056458
F1 Score.............0.147208121827411
Negative Precision...0.913221503103949
Negative Recall......0.997835185452446
Positive Precision...0.794520547945205
Positive Recall......0.0811188811188811

TEST POSITIVE RATIO:    0.0935 (715.0/(715.0+6929.0))
Confusion table
          ||======================
PREDICTED || positive | negative | Recall
TRUTH     ||======================
 positive ||       58 |      657 | 0.0811
 negative ||       15 |    6 914 | 0.9978
          ||======================
Precision ||   0.7945 |   0.9132 |

```
The "Actual" column is a "ground truth label". 
Why am I getting a False prediction when the probability is less than 0.5?
I'm obviously confused.  Please help.

Charles

**Comment by bpstark** [comment]
Hi @CBrauer, The probability here is simply the probability that a given sample is True. If the probability is less than 0.5 then it is more likely to be False then True, thus it is given a False prediction.

Now to go into a little more depth, You are using [Platt scaling](https://en.wikipedia.org/wiki/Platt_scaling) to calculate this probability value. What this means is that we use
![probability function](https://wikimedia.org/api/rest_v1/media/math/render/svg/dd6669e6e04bf09a2d64fb7dd63b8e65d49c34b3)

To create a function which maps the scores which can be of any value to a probability [0,1] that a datapoint is True. The exact values of A and B in the equation are based on the input data, but the idea is to find a function which best represents the entire data set such that for as much of the training data as possible a True example should have a probability greater than 0.5. 

The best way to think about the probability  value is the closer to 0 it is the more likely according to the model that this example is False. The closer to 1 it is the more likely the example is True, and the closer it is to 0.5 the less sure the model is of which category to place the example in. 

Now looking at your precision/recall info, we can see that you have very good precision and recall for your negative samples. However, your positive samples are not nearly as good at recall. Looking at this and the fact that according to the metrics shown less than 10% of your test samples are True, I would assume that your training data is also heavily skewed towards False examples. This makes it much harder to properly classify True examples if there is not enough training data of them. That said I do not have your data set so cannot confirm this.

Hopefully this addresses your question, and if so please close out this issue. If not please let me know, so I can better help.

**Comment by CBrauer** [comment]
Thanks Brian for you comments.

**Comment by CBrauer** [comment]
In my example above, ML.NET produced the following table:
```
First 10 actuals that are true
Actual Predicted     score   probability
------ --------- ----------  -----------
  True     False  -6.288363   0.06514609
  True     False  -7.417452   0.03533126
  True     False  -6.883083    0.0473095
  True      True   2.404708    0.9079463
  True     False  -5.611444   0.09295245
  True      True  0.2836371    0.7465712
  True     False   -1.93054    0.4548629
  True     False  -3.087828    0.3014578
  True      True  0.2553135    0.7435061
  True     False  -5.725763   0.08760489

```
It’s interesting to note that of all the bad predictions, the Score was negative.  
Questions:
1.	What does it mean to have a negative Score value?
2.	In my example above, what is the meaning of “Score”

**Comment by bpstark** [comment]
@CBrauer Score is closely related to probability, in-fact score is used to create probability. The difference is Score Is an unbounded value I.E. It can be any value. Additionally a negative Score simply means a Prediction of False and a positive score means a prediction of True. Thus all bad predictions that you are showing in your snippet are negative, this is because you are showing the first 10 labeled True examples, and a prediction of False must have a negative value. If you were to show labeled False examples that have a bad prediction they would have a positive Score. Probability and Score are just different ways to represent basically the same data. The difference is that probability is Always on the same scale [0,1] where as Score can be any value so it is much harder to understand Score on its own as you don't know how likely a Score is to be correct on its own as there is no reference.  The Platt method transforms the Score data by looking at all present scores, such that it can say how good a prediction is relative to others thus generating a probability from a Score.

Hopefully this addresses your question. If you need further assistance please re-open the issue.

**Comment by rswethashankar92** [comment]
Hi, bpstark your explanation on Platt Calibration was bang on. I need a clarification in an aspect where in fastforest binary classification i need to get the probability for each predicted label. how can i do that ? Please Help.

My Idea is to find the number of trees in agreement with the prediction / Total number of trees in a Fast Forest.
